### PR TITLE
fix: make file name deterministic in parallel emits (fix #4909)

### LINKED ISSUE #4909: Non-deterministic asset file name when multiple asset has same content with a different name
### Rollup Version

3.19.1

### Operating System (or Browser)

stackblitz

### Node Version (if applicable)

_No response_

### Link To Reproduction

https://stackblitz.com/edit/rollup-repro-dju2ff?file=rollup.config.js

### Expected Behaviour

When calling `this.emitFile` with a same content but with different filename, rollup dedupes that and only outputs a single asset.
But because rollup always uses the filename used in the first call, the asset file name is non-deterministic when `this.emitFile` is called in parallel.

For example, Vite outputs CSS files by `this.emitFile` and uses the chunk name as a file name of the CSS file. `renderChunk` is called in parallel and when there's multiple CSS files containing the same content, the build becomes non-deterministic.

I think it would be nice to fix this on rollup side as I guess this can happen by using rollup without Vite.

### Actual Behaviour

Rollup outputs assets with deterministic file names.

One possible solution is to sort all the file names and choose the first one.

Related: https://github.com/vitejs/vite/issues/11911, https://github.com/vitejs/vite/issues/12256

### PR DESCRIPTION
**Pull request by sun0day**
This PR contains:

- [x] bugfix
- [ ] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

List any relevant issue numbers:

### Description

  resolves #4909

### CODE REVIEW

**Comment by sun0day** [comment]
Can you help review this pr? @lukastaegert  Thanks

File: src/utils/FileEmitter.ts [hunk 1]
```diff
@@ -365,14 +365,23 @@ export class FileEmitter {
...
+				source,
+				sourceHash,
+				outputOptions,
+				bundle
+			);
+			// make sure file name deterministic in parallel emits, always use the shorter and smaller file name
+			if (
+				!fileName ||
+				fileName.length > newFileName.length ||
+				(fileName.length === newFileName.length && fileName.localeCompare(newFileName) > 0)
```
Comment (lukastaegert) [inline src/utils/FileEmitter.ts:379]: Are you sure you want to use `localeCompare` here over just using `>`? My worry is that `localCompare` may
- make the output depend on the system language (in edge cases)
- can potentially determine two "different" strings to be the same, in which case again the output would not be deterministic (I think, though not 100% sure)

File: test/chunking-form/samples/emit-file/deduplicate-assets/_config.js [hunk 2]
```diff
@@ -1,16 +1,35 @@
+function asyncify(callback) {
```
Comment (lukastaegert) [inline test/chunking-form/samples/emit-file/deduplicate-assets/_config.js:1]: Instead of making this test random, which could potentially mean that the test is sometimes red and sometimes green but in an indeterministic way, why not work with e.g. fixed numbers as names but in a random order?
e.g. emit 1.txt, 2.txt (same source as 1.txt), 4.txt, 3.txt (same source as 4.txt) and assert that you get 1.txt and 3.txt as output (or 2 and 4, but I hope you see what I mean). Would also make the test nearly 100ms faster 😉

File: src/utils/FileEmitter.ts (hunk 1 above)
Comment (sun0day) [reply src/utils/FileEmitter.ts]: Good point. I will switch to just use `>`

File: test/chunking-form/samples/emit-file/deduplicate-assets/_config.js (hunk 2 above)
Comment (sun0day) [reply test/chunking-form/samples/emit-file/deduplicate-assets/_config.js:1]: I was trying to mock the use case mentioned in #4909. But I am fine with `emitFile` in a random order like you said.

**Review by lukastaegert — APPROVED** [review]
Thanks!

**Comment by lukastaegert** [comment]
Replaced by #4919
